pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,39 +240,192 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+func (c *MCRouteController) ciImpWorker() {
+	for c.processNextClusterInfoImportItem() {
 	}
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
-	obj, quit := c.queue.Get()
+func (c *MCRouteController) processNextClusterInfoImportItem() bool {
+	obj, quit := c.ciImpQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.ciImpQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	ciImp, ok := obj.(string)
+	if !ok {
+		c.ciImpQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.enableWireGuard {
+		if err := c.updatePeers(ciImp); err != nil {
+			c.ciImpQueue.AddRateLimited(ciImp)
+			klog.ErrorS(err, "Error updating WireGuard peers, requeuing", "key", ciImp)
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImpQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.ciImpQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
 	}
 	return true
 }
 
+func (c *MCRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gatewayQueue.Done(obj)
+
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	}
+
+	if c.enableWireGuard && gateway == c.nodeConfig.Name {
+		if err := c.reconcileWireGuard(gateway); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
+	}
+	return true
+}
+
+func (c *MCRouteController) reconcileWireGuard(gateway string) error {
+	if !c.wireGuardInitialized {
+		klog.InfoS("Initializing WireGuard interface and rule")
+		if err := c.initializeWireGuardInterface(); err != nil {
+			return err
+		}
+		if err := c.initHostRoute(); err != nil {
+			return err
+		}
+		c.wireGuardInitialized = true
+		return nil
+	}
+
+	if _, err := c.mcClient.MulticlusterV1alpha1().Gateways(c.namespace).Get(context.TODO(), gateway, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get Gateway %s, error: %s", gateway, err.Error())
+		}
+		klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+		if err := c.deleteHostRoute(); err != nil {
+			return err
+		}
+		c.wireGuardInitialized = false
+	}
+
+	return nil
+}
+
+func (c *MCRouteController) deleteHostRoute() error {
```

Comment:
Question - we do not need any cleanup of WireGuard client, but just need to delete the interface and route?

Reply:
Even, we need not other WG cleanup, we should add a cleanup func in the wireguard package and delete the WG interface there, to be consistent with interface creation in `client.Init()`.